lib/Feature/FeatureModel.cpp

```diff
@@ -14,35 +14,32 @@ void FeatureModel::dump() const {
   llvm::outs() << '\n';
 }
 
-bool FeatureModel::addFeature(std::unique_ptr<Feature> NewFeature) {
-  // TODO(s9latimm): check consistency
-  auto FeatureName = std::string(NewFeature->getName());
-  if (!Features.try_emplace(FeatureName, std::move(NewFeature)).second) {
-    return false;
-  }
-  auto *Value = Features[FeatureName].get();
-  for (auto *Child : Value->children()) {
-    Child->setParent(*Value);
+Feature *FeatureModel::addFeature(std::unique_ptr<Feature> NewFeature) {
+  auto PosInsertedFeature = Features.try_emplace(
+      std::string(NewFeature->getName()), std::move(NewFeature));
+  if (!PosInsertedFeature.second) {
+    return nullptr;
   }
-  if (Value->isRoot()) {
-    if (*Root->getParentFeature() == *Value) {
-      Root = Value;
-    } else {
-      Value->setParent(Root);
-    }
-  } else {
-    Value->getParent()->addEdge(Value);
+  auto *InsertedFeature = PosInsertedFeature.first->getValue().get();
+  assert(InsertedFeature);
+  OrderedFeatures.insert(InsertedFeature);
+  return InsertedFeature;
+}
+
+void FeatureModel::removeFeature(Feature &F) {
+  if (&F == Root) {
+    Root = nullptr;
   }
-  OrderedFeatures.insert(Value);
-  return true;
+  OrderedFeatures.remove(&F);
+  Features.erase(F.getName());
```

Comment:
Should we also ensure that the child feature no longer use this feature as parent? Or is this already the case?

Reply:
Nope, the FM just offers private modifier methods that exactly change what requested, however, these are all private and only accessible to Modifications (after we refactored the builder). The Modification itself then guarantees and is tested to ensure that (e.g., AddFeature) does the correct thing, setting parent/child relation ships etc..